crates/dbsp/src/circuit/circuit_builder.rs

```diff
@@ -4935,22 +5843,357 @@ impl CircuitHandle {
+
+        let mut participate_in_backfill_new = BTreeSet::new();
+
+        let mut replay_streams = BTreeMap::new();
+
+        for (stream_id, mut node_id) in inputs.into_iter() {
+            // println!("replay needed for ({stream_id}, {node_id})");
+
+            // Add all ancestors of `participate_in_backfill_new` to the `participate_in_backfill` set, except streams
+            // that can be replayed from a different node.
+            if let Some(stream_id) = stream_id {
+                if let Some(replay_source) = self.circuit.get_replay_source(stream_id) {
+                    // If the replay source is itself in the need_backfill set (i.e., it's an integral without
+                    // a checkpoint), it cannot be used for replay.
+                    if !need_backfill.contains(&replay_source.local_node_id()) {
+                        replay_streams.insert(stream_id, replay_source.clone());
+                        // trace!(
+                        //     "worker {}: Replacing node_id {node_id} with replay source {}",
+                        //     Runtime::worker_index(),
+                        //     replay_source.origin_node_id()
+                        // );
+
+                        // Replace the node_id with the replay source.
+                        node_id = replay_source.local_node_id();
+                    }
+                }
+            }
+
+            if !participate_in_backfill.contains(&node_id) {
+                // println!("Adding {gid} to participate_in_backfill via {stream_id:?}");
+                participate_in_backfill.insert(node_id);
+                participate_in_backfill_new.insert(node_id);
+            }
+        }
+
+        // Connect `replay_streams` to all operators that consume the original stream.
+        for (original_stream, replay_stream) in replay_streams.into_iter() {
+            replay_sources
+                .entry(replay_stream.local_node_id())
+                .or_insert_with(|| {
+                    self.circuit
+                        .add_replay_edges(original_stream, replay_stream.as_ref());
+                    replay_stream.stream_id()
+                });
+        }
+
+        Ok(participate_in_backfill_new)
+    }
+
+    /// Returns `true` if all replay sources have completed their replay.
+    pub fn is_replay_complete(&self) -> bool {
+        let Some(replay_info) = self.replay_info.as_ref() else {
+            return true;
+        };
+
+        replay_info.replay_sources.keys().all(|node_id| {
+            self.circuit
+                .map_local_node_mut(*node_id, &mut |node| node.is_replay_complete())
+        })
+    }
+
+    /// Finalize the replay phase of the circuit.
+    ///
+    /// * Notify all replay sources to go back to normal operation.
+    /// * Delete all replay streams.
+    /// * Prepare the scheduler to run the full circuit.
+    pub fn complete_replay(&mut self) -> Result<(), DbspError> {
+        // info!("Replay complete");
```

Comment:
this also looks like it should be left in, perhaps even with a higher logging level